![Genomic-disparity](https://github.com/nhm-herpetology/Genomic-disparity/blob/main/Genomic-disparity.jpg)

## An approach for exploring chromosome variation in high-quality genome assemblies

Genomic architecture has played a key role in the evolution of biodiversity. Structural comparisons of genome sequences have informed the study of supergenes, sex chromosomes, and some of the earliest divergences in the tree of life. **Genomic Disparity Analysis** is a computationally non-intensive approach, based on geometric morphometrics, that captures disparity scores across conserved chromosomal regions using single-copy, conserved genomic regions as landmarks (e.g. ultraconserved elements [UCEs]). The method allows users to visualize structural variation in multivariate space using PCA for many species or individuals at the same time.

>Citation:

**Mohan, A.V., Goswami, A., Streicher J.W. 2024. Landmark-based estimates of genomic disparity. Under Revision.**

<br/> 

## Exploring genomic variation with UCEs: a deep phylogeny tutorial
Using Genomic Disparity Analysis to study structural evolution on multiple chromosomes from placental mammals.

>Click [here](https://github.com/nhm-herpetology/genomic-disparity/tree/main/Mammal-uces/README.md) to access tutorial.

<br/>

## Detecting chromosomal inversions with UCEs: a speciation tutorial
Using Genomic Disparity Analysis to explore variation in landmark placement on a single chromosome in the _Drosophila virilis_ group.

>Click [here](https://github.com/nhm-herpetology/genomic-disparity/tree/main/Drosophila-uces/README.md) to access tutorial. 

<br/>




